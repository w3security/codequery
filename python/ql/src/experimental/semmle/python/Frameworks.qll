/**
 * Helper file that imports all framework modeling.
 */

private import experimental.semmle.python.frameworks.Stdlib
private import experimental.semmle.python.frameworks.Flask
private import experimental.semmle.python.frameworks.Django
private import experimental.semmle.python.frameworks.Werkzeug
private import experimental.semmle.python.frameworks.LDAP
private import experimental.semmle.python.frameworks.NoSQL
private import experimental.semmle.python.frameworks.Log
