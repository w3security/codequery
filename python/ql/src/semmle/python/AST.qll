import python

/** Syntactic node (Class, Function, Module, Expr, Stmt or Comprehension) corresponding to a flow node */
abstract class AstNode extends AstNode_ {

    /** Gets the scope that this node occurs in */
    abstract Scope getScope();

    /** Gets a flow node corresponding directly to this node.
     * NOTE: For some statements and other purely syntactic elements,
     * there may not be a `ControlFlowNode` */
    ControlFlowNode getAFlowNode() {
        py_flow_bb_node(result, this, _, _)
    }

    /** Gets the location for this AST node */
    Location getLocation() {
        none()
    }

    /** Whether this syntactic element is artificial, that is it is generated 
     *  by the compiler and is not present in the source */
    predicate isArtificial() {
        none()
    }

    /** Gets a child node of this node in the AST. This predicate exists to aid exploration of the AST
      * and other experiments. The child-parent relation may not be meaningful. 
      * For a more meaningful relation in terms of dependency use 
      * Expr.getASubExpression(), Stmt.getASubStatement(), Stmt.getASubExpression() or
      * Scope.getAStmt().
      */
    abstract AstNode getAChildNode();

    /** Gets the parent node of this node in the AST. This predicate exists to aid exploration of the AST
      * and other experiments. The child-parent relation may not be meaningful. 
      * For a more meaningful relation in terms of dependency use 
      * Expr.getASubExpression(), Stmt.getASubStatement(), Stmt.getASubExpression() or
      * Scope.getAStmt() applied to the parent.
      */
    AstNode getParentNode() {
        result.getAChildNode() = this
    }

    /** Whether this contains `inner` syntactically */
    predicate contains(AstNode inner) {
        this.getAChildNode+() = inner
    }

    /** Whether this contains `inner` syntactically and `inner` has the same scope as `this` */
    predicate containsInScope(AstNode inner) {
        this.contains(inner) and
        this.getScope() = inner.getScope() and
        not inner instanceof Scope
    }

}
