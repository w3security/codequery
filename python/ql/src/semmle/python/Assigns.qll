/** 
 * In order to handle data flow and other analyses efficiently the extractor transforms various statements which perform binding in assignments.
 * These classes provide a wrapper to provide a more 'natural' interface to the syntactic elements transformed to assignments.
 */

import python


/** An assignment statement */
class AssignStmt extends Assign {

    AssignStmt() {
        not this instanceof FunctionDef and not this instanceof ClassDef
    }

    override string toString() {
        result = "AssignStmt"
    }
}
