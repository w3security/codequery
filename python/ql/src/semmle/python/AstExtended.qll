import python

/* Parents */

/** Internal implementation class */
library class FunctionParent extends FunctionParent_ {

}

/** Internal implementation class */
library class ArgumentsParent extends ArgumentsParent_ {

}

/** Internal implementation class */
library class ExprListParent extends ExprListParent_ {

}

/** Internal implementation class */
library class ExprContextParent extends ExprContextParent_ {

}

/** Internal implementation class */
library class StmtListParent extends StmtListParent_ {

}

/** Internal implementation class */
library class StrListParent extends StrListParent_ {

}

/** Internal implementation class */
library class ExprParent extends ExprParent_ {

}

library class DictItem extends DictItem_, AstNode {

    override string toString() {
        result = DictItem_.super.toString()
    }

    override AstNode getAChildNode() { none() }

    override Scope getScope() { none() }

}

/** A comprehension part, the 'for a in seq' part of  [ a * a for a in seq ] */
class Comprehension extends Comprehension_, AstNode {

    /** Gets the scope of this comprehension */
    override Scope getScope() {
        /* Comprehensions exists only in Python 2 list comprehensions, so their scope is that of the list comp. */
        exists(ListComp l |
            this = l.getAGenerator() |
            result = l.getScope() 
        )
    }

    override string toString() {
        result = "Comprehension"
    }

    override Location getLocation() {
        result = Comprehension_.super.getLocation()
    }

    override AstNode getAChildNode() {
        result = this.getASubExpression()
    }

    Expr getASubExpression() {
        result = this.getIter() or
        result = this.getAnIf() or
        result = this.getTarget()
    }

}

class BytesOrStr extends BytesOrStr_ { 

}

/** Part of a string literal formed by implicit concatenation.
 * For example the string literal "abc" expressed in the source as `"a" "b" "c"`
 * would be composed of three `StringPart`s.
 * 
 */
class StringPart extends StringPart_, AstNode {

    override Scope getScope() {
        exists(Bytes b | this = b.getAnImplicitlyConcatenatedPart() | result = b.getScope())
        or
        exists(Unicode u | this = u.getAnImplicitlyConcatenatedPart() | result = u.getScope())
    }

    override AstNode getAChildNode() {
        none()
    }

    override string toString() {
        result = StringPart_.super.toString()
    }

    override Location getLocation() {
        result = StringPart_.super.getLocation()
    }

}

class StringPartList extends StringPartList_ {

}

