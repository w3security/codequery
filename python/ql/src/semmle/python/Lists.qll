import python

/** A parameter list */
class ParameterList extends @py_parameter_list {

    Function getParent() {
        py_parameter_lists(this, result)
    }

    /** Gets a parameter */
    Parameter getAnItem() {
        /* Item can be a Name or a Tuple, both of which are expressions */
        py_exprs(result, _, this, _)
    }

    /** Gets the nth parameter */
    Parameter getItem(int index) {
        /* Item can be a Name or a Tuple, both of which are expressions */
        py_exprs(result, _, this, index)
    }

    string toString() {
        result = "ParameterList"
    }
}

/** A list of Comprehensions (for generating parts of a set, list or dictionary comprehension) */
class ComprehensionList extends ComprehensionList_ {

}

/** A list of expressions */
class ExprList extends ExprList_ {

}


library class DictItemList extends DictItemList_ {

}

library class DictItemListParent extends DictItemListParent_ {

}

/** A list of strings (the primitive type string not Bytes or Unicode) */
class StringList extends StringList_ {

}

/** A list of aliases in an import statement */
class AliasList extends AliasList_ {

}

