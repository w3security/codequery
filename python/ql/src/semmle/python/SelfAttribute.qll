/** Utilities to support queries about instance attribute accesses of
 * the form `self.attr`.
 */

import python
private import semmle.python.pointsto.PointsTo
private import semmle.python.pointsto.Filters

/** An attribute access where the left hand side of the attribute expression 
  * is `self`.
  */
class SelfAttribute extends Attribute {

    SelfAttribute() {
        self_attribute(this, _)
    }

    Class getClass() {
        self_attribute(this, result)
    }

}

/** Whether variable 'self' is the self variable in method 'method' */
private predicate self_variable(Function method, Variable self) {
    self.isParameter() and
    method.isMethod() and
    method.getArg(0).asName() = self.getAnAccess()
}

/** Whether attribute is an access of the form `self.attr` in the body of the class 'cls' */
private predicate self_attribute(Attribute attr, Class cls) {
    exists(Function f, Variable self |
        self_variable(f, self) |
        self.getAnAccess() = attr.getObject() and
        cls = f.getScope+()
    )
}

/** Helper class for UndefinedClassAttribute.ql &amp; MaybeUndefinedClassAttribute.ql */
class SelfAttributeRead extends SelfAttribute {

    SelfAttributeRead() {
        this.getCtx() instanceof Load and
        /* Be stricter for loads. 
         * We want to generous as to what is defined (ie stores),
         * but strict as to what needs to be defined (ie loads).
         */
        exists(ClassObject cls, FunctionObject func |
            cls.declaredAttribute(_) = func |
            func.getFunction() = this.getScope() and
            cls.getPyClass() = this.getClass()
        )
    }

    predicate guardedByHasattr() {
        exists(Variable var, ControlFlowNode n |
            var.getAUse() = this.getObject().getAFlowNode() and
            hasattr(n, var.getAUse(), this.getName()) and
            n.strictlyDominates(this.getAFlowNode())
        )
    }

    pragma [noinline] predicate locallyDefined() {
        exists(SelfAttributeStore store |
            this.getName() = store.getName() and 
            this.getScope() = store.getScope() |
            store.getAFlowNode().strictlyDominates(this.getAFlowNode())
        )
    }

}

class SelfAttributeStore extends SelfAttribute {

    SelfAttributeStore() {
        this.getCtx() instanceof Store
    }

    Expr getAssignedValue() {
        exists(Assign a | a.getATarget() = this |
            result = a.getValue()
        )
    }

}

private Object object_getattribute() {
    py_cmembers_versioned(theObjectType(), "__getattribute__", result, major_version().toString())
}

/** Helper class for UndefinedClassAttribute.ql and MaybeUndefinedClassAttribute.ql */
class CheckClass extends ClassObject {

    private predicate ofInterest() {
        not this.unknowableAttributes() and
        not this.getPyClass().isProbableMixin() and
        this.getPyClass().isPublic() and
        not this.getPyClass().getScope() instanceof Function and
        not this.probablyAbstract() and
        not this.declaresAttribute("__new__") and
        not this.selfDictAssigns() and
        not this.lookupAttribute("__getattribute__") != object_getattribute() and
        not this.hasAttribute("__getattr__") and
        not this.selfSetattr() and
        /* If class overrides object.__init__, but we can't resolve it to a Python function then give up */
        forall(ClassObject sup |
            sup = this.getAnImproperSuperType() and
            sup.declaresAttribute("__init__") and
            not sup = theObjectType() |
            sup.declaredAttribute("__init__") instanceof PyFunctionObject
        )
    }

    predicate alwaysDefines(string name) {
        auto_name(name) or
        this.hasAttribute(name) or
        this.getAnImproperSuperType().assignedInInit(name) or
        this.getMetaClass().assignedInInit(name)
    }

    predicate sometimesDefines(string name) {
        this.alwaysDefines(name) or
        exists(SelfAttributeStore sa | 
            sa.getScope().getScope+() = this.getAnImproperSuperType().getPyClass() |
            name = sa.getName()
        )
    }

    private predicate selfDictAssigns() {
        exists(Assign a, SelfAttributeRead self_dict, Subscript sub | 
            self_dict.getName() = "__dict__" and
            ( 
              self_dict = sub.getObject() 
              or
              /* Indirect assignment via temporary variable */
              exists(SsaVariable v | 
                  v.getAUse() = sub.getObject().getAFlowNode() and 
                  v.getDefinition().(DefinitionNode).getValue() = self_dict.getAFlowNode()
              )
            ) and
            a.getATarget() = sub and
            exists(FunctionObject meth | meth = this.lookupAttribute(_) and a.getScope() = meth.getFunction())
        )
    }

    pragma [nomagic]
    private predicate monkeyPatched(string name) {
        exists(Attribute a |
             a.getCtx() instanceof Store and
             PointsTo::points_to(a.getObject().getAFlowNode(), _, this, _, _) and a.getName() = name
        )
    }

    private predicate selfSetattr() {
      exists(Call c, Name setattr, Name self, Function method |
          ( method.getScope() = this.getPyClass() or 
            method.getScope() = this.getASuperType().getPyClass()
          ) and
          c.getScope() = method and
          c.getFunc() = setattr and
          setattr.getId() = "setattr" and
          c.getArg(0) = self and
          self.getId() = "self"
      )
    }

  predicate interestingUndefined(SelfAttributeRead a) {
      exists(string name | name = a.getName() |
          interestingContext(a, name) and
          not this.definedInBlock(a.getAFlowNode().getBasicBlock(), name)
      )
  }

  private predicate interestingContext(SelfAttributeRead a, string name) {
      name = a.getName() and
      this.ofInterest() and
      this.getPyClass() = a.getScope().getScope() and
      not a.locallyDefined() and
      not a.guardedByHasattr() and
      a.getScope().isPublic() and
      not this.monkeyPatched(name) and
      not attribute_assigned_in_method(lookupAttribute("setUp"), name)
  }

  private predicate probablyAbstract() {
      this.getName().matches("Abstract%")
      or
      this.isAbstract()
  }

  private pragma[nomagic] predicate definitionInBlock(BasicBlock b, string name) {
      exists(SelfAttributeStore sa | 
          sa.getAFlowNode().getBasicBlock() = b and sa.getName() = name and sa.getClass() = this.getPyClass()
      )
      or
      exists(FunctionObject method | this.lookupAttribute(_) = method |
          attribute_assigned_in_method(method, name) and
          b = method.getACall().getBasicBlock()
      )
  }

  private pragma[nomagic] predicate definedInBlock(BasicBlock b, string name) {
      // manual specialisation: this is only called from interestingUndefined,
      // so we can push the context in from there, which must apply to a
      // SelfAttributeRead in the same scope
      exists(SelfAttributeRead a | 
          a.getScope() = b.getScope() and name = a.getName() |
          interestingContext(a, name)
      )
      and
      this.definitionInBlock(b, name)
      or
      exists(BasicBlock prev | this.definedInBlock(prev, name) and prev.getASuccessor() = b)
  }

}

private predicate attr_assigned_in_method_arg_n(FunctionObject method, string name, int n) {
    exists(SsaVariable param |
        method.getFunction().getArg(n).asName() = param.getDefinition().getNode()
        |
        exists(AttrNode attr | 
            attr.getObject(name) = param.getAUse() and
            attr.isStore()
        )
        or
        exists(CallNode call, FunctionObject callee, int m |
            callee.getArgumentForCall(call, m) = param.getAUse() and
            attr_assigned_in_method_arg_n(callee, name, m)
        )
    )
}

predicate attribute_assigned_in_method(FunctionObject method, string name) {
    attr_assigned_in_method_arg_n(method, name, 0)
}

private predicate auto_name(string name) {
  name = "__class__" or name = "__dict__"
}
