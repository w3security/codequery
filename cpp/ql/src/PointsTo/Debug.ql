/**
 * @name Debug - find out what a particular function-pointer points to
 * @description Query to help investigate mysterious results with ReturnStackAllocatedObject
 * @kind table
 * @id cpp/points-to/debug
 */
import cpp
import semmle.code.cpp.pointsto.PointsTo

class FieldAccessPT extends PointsToExpr
{ override predicate interesting() { this instanceof FieldAccess } }

from Function outer, FieldAccessPT fa
where outer.hasName("rtLnDeliverableMayContainDividends")
  and fa.(FieldAccess).getTarget().hasName("pfFunction")
  and fa.getEnclosingFunction() = outer
select fa, fa.pointsTo()
