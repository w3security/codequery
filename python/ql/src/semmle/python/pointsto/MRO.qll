/** Classes and predicates for computing the Method Resolution Order (MRO) of classes.
 * Supports both old-style (diamond) inheritance and new-style (C3 linearization) inheritance.
 */

/*
 * Implementation of the C3 linearization algorithm.
 * See https://en.wikipedia.org/wiki/C3_linearization
 *
 * The key operation is merge, which takes a list of lists and produces a list.
 * We implement it as the method `ClassListList.merge()`
 *
 * To support that we need to determine the best candidate to extract from a list of lists,
 * implemented as `ClassListList.bestMergeCandidate()`
 *
 * The following code is designed to implement those operations
 * without negation and as efficiently as possible.
 */

import python
import semmle.python.pointsto.PointsTo

cached private newtype TClassList = Empty()
    or
    Cons(ClassObject head, TClassList tail) {
        required_cons(head, tail)
    }

/* Keep ClassList finite and as small as possible */
private predicate required_cons(ClassObject head, ClassList tail) {
    tail = merge_of_linearization_of_bases(head)
    or
    exists(ClassObject cls, int n |
        head = cls.getBaseType(n) and tail = bases(cls, n+1)
    )
    or
    head = theObjectType() and tail = Empty()
    or
    reverse_step(_, Cons(head, _), tail)
    or
    exists(ClassListList list |
        merge_step(tail, list, _) and
        head = list.bestMergeCandidate()
    )
    or
    exists(ClassList list, int n |
        n = list.firstIndex(head) and
        tail = list.deduplicate(n+1)
    )
    or
    exists(ClassListList list, int n |
        head = list.getHead().getItem(n) and
        tail = flatten_list(list, n+1)
    )
    or
    tail = list_old_style_base_mros(head).flatten()
}

/** A list of classes, used to represent the MRO of a class */
class ClassList extends TClassList {

    string toString() {
        result = "[" + this.contents() + "]"
    }

    string contents() {
        this = Empty() and result = ""
        or
        exists(ClassObject head |
            head = this.getHead() |
            this.getTail() = Empty() and result = head.getName()
            or
            this.getTail() != Empty() and result = head.getName() + ", " + this.getTail().contents()
        )
    }

    int length() {
        this = Empty() and result = 0
        or
        result = this.getTail().length() + 1
    }

    ClassObject getHead() {
        this = Cons(result, _)
    }

    ClassList getTail() {
        this = Cons(_, result)
    }

    ClassObject getItem(int n) {
        n = 0 and  result = this.getHead()
        or
        result = this.getTail().getItem(n-1)
    }

    pragma [inline]
    ClassList removeHead(ClassObject cls) {
        this.getHead() = cls and result = this.getTail()
        or
        this.getHead() != cls and result = this
        or
        this = Empty() and result = Empty()
    }

    predicate legalMergeHead(ClassObject cls) {
        this.getTail().doesNotContain(cls)
        or
        this = Empty()
    }

    /** Use negative formulation for efficiency */
    predicate contains(ClassObject cls) {
        cls = this.getHead()
        or
        this.getTail().contains(cls)
    }

    /** Use negative formulation to avoid negative recursion */
    predicate doesNotContain(ClassObject cls) {
        this.relevantForContains(cls) and 
        cls != this.getHead() and 
        this.getTail().doesNotContain(cls)
        or
        this = Empty()
    }

    private predicate relevantForContains(ClassObject cls) {
        exists(ClassListList list |
            list.getItem(_).getHead() = cls and
            list.getItem(_) = this
        )
        or
        exists(ClassList l |
            l.relevantForContains(cls) and
            this = l.getTail()
        )
    }

    ClassObject findDeclaringClass(string name) {
        exists(ClassObject head |
            head = this.getHead() and
            not head = theUnknownType() |
            if head.declaresAttribute(name) then
                result = head
            else
                result = this.getTail().findDeclaringClass(name)
        )
    }

    Object lookup(string name) {
        exists(ClassObject head |
            head = this.getHead() and
            not head = theUnknownType() |
            if head.declaresAttribute(name) then
                result = head.declaredAttribute(name)
            else
                result = this.getTail().lookup(name)
        )
    }

    predicate declares(string name) {
        this.getHead().declaresAttribute(name)
        or
        this.getTail().declares(name)
    }

    ClassList startingAt(ClassObject cls) {
        exists(ClassObject head |
           head = this.getHead() |
            if head = cls then
                result = this
            else
                result = this.getTail().startingAt(cls)
        )
    }

    ClassList deduplicate() {
        result = this.deduplicate(0)
    }

    /* Helpers for `deduplicate()` */

    int firstIndex(ClassObject cls) {
        result = this.firstIndex(cls, 0)
    }

    /* Helper for firstIndex(cls), getting the first index of `cls` where result >= n */
    private int firstIndex(ClassObject cls, int n) {
        this.getItem(n) = cls and result = n
        or
        this.getItem(n) != cls and result = this.firstIndex(cls, n+1)
    }

    /** Holds if the class at `n` is a duplicate of an earlier position. */
    private predicate duplicate(int n) {
        exists(ClassObject cls |
            cls = this.getItem(n) and this.firstIndex(cls) < n
        )
    }

    /** Gets a class list which is the de-duplicated form of the list containing elements of
     * this list from `n` onwards.
     */
    ClassList deduplicate(int n) {
        n = this.length() and result = Empty()
        or
        this.duplicate(n) and result = this.deduplicate(n+1)
        or
        exists(ClassObject cls |
            n = this.firstIndex(cls) and
            result = Cons(cls, this.deduplicate(n+1))
        )
    }

    predicate isEmpty() {
        this = Empty()
    }

    ClassList reverse() {
        reverse_step(this, Empty(), result)
    }
}

private newtype TClassListList =
    EmptyList() or
    ConsList(TClassList head, TClassListList tail) {
        required_list(head, tail)
    }

/* Keep ClassListList finite and as small as possible */
private predicate required_list(ClassList head, ClassListList tail) {
    any(ClassListList x).removedClassParts(_, head, tail, _)
    or
    head = bases(_) and tail = EmptyList()
    or
    exists(ClassObject cls, int n |
        head = new_style_mro(cls.getBaseType(n)) and
        tail = list_of_linearization_of_bases_plus_bases(cls, n+1)
    )
    or
    exists(ClassObject cls, int n |
        head = old_style_mro(cls.getBaseType(n)) and
        tail = list_old_style_base_mros(cls, n+1)
    )
}

private class ClassListList extends TClassListList {

    string toString() {
        result = "[" + this.contents() + "]"
    }

    string contents() {
        this = EmptyList() and result = ""
        or
        exists(ClassList head |
            head = this.getHead() |
            this.getTail() = EmptyList() and result = head.toString()
            or
            this.getTail() != EmptyList() and result = head.toString() + ", " + this.getTail().contents()
        )
    }

    int length() {
        this = EmptyList() and result = 0
        or
        result = this.getTail().length() + 1
    }

    ClassList getHead() {
        this = ConsList(result, _)
    }

    ClassListList getTail() {
        this = ConsList(_, result)
    }

    ClassList getItem(int n) {
        n = 0 and result = this.getHead()
        or
        result = this.getTail().getItem(n-1)
    }

    private ClassObject getAHead() {
        result = this.getHead().getHead()
        or
        result = this.getTail().getAHead()
    }

    pragma [nomagic]
    ClassList merge() {
        exists(ClassList reversed |
            merge_step(reversed, EmptyList(), this) and
            result = reversed.reverse()
        )
        or
        this = EmptyList() and result = Empty()
    }

    /* Join ordering helper */
    pragma [noinline]
    predicate removedClassParts(ClassObject cls, ClassList removed_head, ClassListList removed_tail, int n) {
        cls = this.bestMergeCandidate() and n = this.length()-1 and
        removed_head = this.getItem(n).removeHead(cls) and removed_tail = EmptyList()
        or
        exists(ClassList prev_head, ClassListList prev_tail |
            this.removedClassParts(cls, prev_head, prev_tail, n+1) and
            removed_head = this.getItem(n).removeHead(cls) and
            removed_tail = ConsList(prev_head, prev_tail)
        )
    }

    ClassListList remove(ClassObject cls) {
        exists(ClassList removed_head, ClassListList removed_tail |
            this.removedClassParts(cls, removed_head, removed_tail, 0) and
            result = ConsList(removed_head, removed_tail)
        )
        or
        this = EmptyList() and result = EmptyList()
    }

    predicate legalMergeCandidate(ClassObject cls, int n) {
        cls = this.getAHead() and n = this.length()
        or
        this.getItem(n).legalMergeHead(cls) and
        this.legalMergeCandidate(cls, n+1)
    }

    predicate legalMergeCandidate(ClassObject cls) {
        this.legalMergeCandidate(cls, 0)
    }

    predicate illegalMergeCandidate(ClassObject cls) {
        cls = this.getAHead() and
        this.getItem(_).getTail().contains(cls)
    }

    ClassObject bestMergeCandidate(int n) {
        exists(ClassObject head |
            head = this.getItem(n).getHead()
            |
            legalMergeCandidate(head) and result = head
            or
            illegalMergeCandidate(head) and result = this.bestMergeCandidate(n+1)
        )
    }

    ClassObject bestMergeCandidate() {
        result = this.bestMergeCandidate(0)
    }

    /** Gets a ClassList representing the this list of list flattened into a single list.
     * Used for old-style MRO computation.
     */
    ClassList flatten() {
        this = EmptyList() and result = Empty()
        or
        result = flatten_list(this, 0)
    }

}

private ClassList flatten_list(ClassListList list, int n) {
    need_flattening(list) and
    exists(ClassList head, ClassListList tail |
        list = ConsList(head, tail)
        |
        n = head.length() and result = tail.flatten()
        or
        result = Cons(head.getItem(n), flatten_list(list, n+1))
    )
}

/* Restrict flattening to those lists that need to be flattened */
private predicate need_flattening(ClassListList list) {
    list = list_old_style_base_mros(_)
    or
    exists(ClassListList toflatten |
        need_flattening(toflatten) and
        list = toflatten.getTail()
    )
}

private ClassList bases(ClassObject cls) {
    result = bases(cls, 0)
}

private ClassList bases(ClassObject cls, int n) {
    result = Cons(cls.getBaseType(n), bases(cls, n+1))
    or
    result = Empty() and n = PointsTo::Types::class_base_count(cls)
}

private ClassListList list_of_linearization_of_bases_plus_bases(ClassObject cls) {
    result = list_of_linearization_of_bases_plus_bases(cls, 0)
}

private ClassListList list_of_linearization_of_bases_plus_bases(ClassObject cls, int n) {
    result = ConsList(bases(cls), EmptyList()) and n = PointsTo::Types::class_base_count(cls)
    or
    exists(ClassListList partial |
        partial = list_of_linearization_of_bases_plus_bases(cls, n+1) and
        result = ConsList(new_style_mro(cls.getBaseType(n)), partial)
    )
}

private ClassList merge_of_linearization_of_bases(ClassObject cls) {
    result = list_of_linearization_of_bases_plus_bases(cls).merge()
}

cached ClassList new_style_mro(ClassObject cls) {
    cls = theObjectType() and result = Cons(cls, Empty())
    or
    result = Cons(cls, merge_of_linearization_of_bases(cls))
}

cached ClassList old_style_mro(ClassObject cls) {
    PointsTo::Types::is_new_style_bool(cls) = false and
    result = Cons(cls, list_old_style_base_mros(cls).flatten()).(ClassList).deduplicate()
}

private ClassListList list_old_style_base_mros(ClassObject cls) {
    result = list_old_style_base_mros(cls, 0)
}

pragma [nomagic]
private ClassListList list_old_style_base_mros(ClassObject cls, int n) {
    n = PointsTo::Types::class_base_count(cls) and result = EmptyList()
    or
    result = ConsList(old_style_mro(cls.getBaseType(n)), list_old_style_base_mros(cls, n+1))
}

/** Holds if the pair `reversed_mro`, `remaining_list` represents a step in the C3 merge operation
 * of computing the C3 linearization of `original`.
 */
private predicate merge_step(ClassList reversed_mro, ClassListList remaining_list, ClassListList original) {
    remaining_list = list_of_linearization_of_bases_plus_bases(_) and reversed_mro = Empty() and remaining_list = original
    or
    /* Removes the best merge candidate from `remaining_list` and prepends it to `reversed_mro` */
    exists(ClassObject head, ClassList prev_reverse_mro, ClassListList prev_list |
        merge_step(prev_reverse_mro, prev_list, original) and
        head = prev_list.bestMergeCandidate() and
        reversed_mro = Cons(head, prev_reverse_mro) and
        remaining_list = prev_list.remove(head)
    )
    or
    merge_step(reversed_mro, ConsList(Empty(), remaining_list), original)
}

/* Helpers for `ClassList.reverse()` */

private predicate needs_reversing(ClassList lst) {
    merge_step(lst, EmptyList(), _)
    or
    lst = Empty()
}

private predicate reverse_step(ClassList lst, ClassList remainder, ClassList reversed) {
    needs_reversing(lst) and remainder = lst and reversed = Empty()
    or
    exists(ClassObject head, ClassList tail |
        reversed = Cons(head, tail) and
        reverse_step(lst, Cons(head, remainder), tail)
    )
}

