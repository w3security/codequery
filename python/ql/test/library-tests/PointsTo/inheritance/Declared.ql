
import python
import semmle.python.pointsto.PointsTo

from ClassObject cls, string name, PyFunctionObject f
where PointsTo::Types::class_declared_attribute(cls, name, f, _, _)
select cls.toString(), name, f.toString(), f.getFunction().getLocation().getStartLine()
