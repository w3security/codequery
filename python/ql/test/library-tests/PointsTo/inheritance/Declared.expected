| class Base | meth | Function meth | 3 |
| class Derived1 | meth | Function meth | 8 |
| class Derived2 | meth | Function meth | 13 |
| class Derived4 | meth | Function meth | 21 |
| class Derived5 | meth | Function meth | 26 |
| class Derived6 | meth | Function meth | 31 |
| class Missing1 | a | Function a | 49 |
| class Missing2 | b | Function b | 53 |
| class Missing3 | c | Function c | 57 |
| class Wrong1 | meth | Function meth | 37 |
| class Wrong2 | meth | Function meth | 42 |
